src/python/pants/backend/python/tasks2/resolve_requirements_task_base.py

```diff
@@ -43,14 +52,25 @@ def resolve_requirements(self, req_libs):
       interpreter = self.context.products.get_data(PythonInterpreter)
       path = os.path.realpath(os.path.join(self.workdir, str(interpreter.identity), target_set_id))
 
+      # Gather invalid targets so we can check whether the requirements pex should be
+      # rebuilt due to changes to PythonDistribution targets.
+      invalid_targets = [vt.target for vt in invalidation_check.invalid_vts]
       # Note that we check for the existence of the directory, instead of for invalid_vts,
       # to cover the empty case.
-      if not os.path.isdir(path):
+      if not os.path.isdir(path) or targets_are_invalid(python_dist_targets, invalid_targets):
         with safe_concurrent_creation(path) as safe_path:
           self._build_requirements_pex(interpreter, safe_path, req_libs)
     return PEX(path, interpreter=interpreter)
 
   def _build_requirements_pex(self, interpreter, path, req_libs):
     builder = PEXBuilder(path=path, interpreter=interpreter, copy=True)
     dump_requirements(builder, interpreter, req_libs, self.context.log)
+    # Dump built python distributions, if any, into the requirements pex.
+    # NB: If a python_dist depends on a requirement X and another target in play requires
+    # an incompatible version of X, this will cause the pex resolver to throw an incompatiblity
+    # error and Pants will abort the build.
```

Comment:
These sound like fair concerns. Let me ask a couple clarifying questions to make sure I understand them correctly. I'm not entirely clear so this might be easier to clarify over chat / hangouts. Lmk if that's viable for you.

1) I assume disallowing `dependencies=` would be mainly a #simplify measure, but it wouldn't cover the test case that you mention?
2) Re: test case are you referring to a `python_dist` with a setup.py that has:

```
setup(
  name='pytorch',
  version='1.2.3'
)
```

and a target that depends on this `python_dist` and another target that depends on `1.2.4` transitively like:

```
python_binary(
  name='main',
  source='main.py',
  dependencies=[
    ':main2',
    ':pytorch',
  ]
)

python_library(
  name='main2',
  source='main2.py',
  requirements=[
    'python_requirement_library(pytorch==1.2.4)',
  ]
)

3) Hope I'm wrong here, but it seems like resolving transitive deps is a pretty complex/nontrivial undertaking. I understand the need to #BeRigorous, but I'm wondering if we can balance it with "Iterative, team-based development"? What if "this feels like a feature customers will need from day 1" weren't true for our internal customer that requested this feature? Assuming we address all other concerns, is there a path we can take to unblock our customer now and tackle transitive deps in a separate pull request?

Reply:
1. correct, two separate things there.

2. that would be exactly it for the conflict case. for transitive dep resolution, the idea would be a `python_binary` that depended on a `python_dist` with a `setup.py` that declared a `install_requires=['some_transitive_dep']` - which would be transitively imported from the `python_binary`'s `main.py` during a `test`/`run`. I believe currently this would fail on the import of `some_transitive_dep` due to no resolver treatment for `setup.py` declared deps.

3. hmm, I don't get the sense it'd be super complex. the first bit would be extracting the injectable requirement spec from the built wheel or `setup.py`, but you can probably get away with simple post-build wheel filename parsing for that (and then e.g. `'=='.join([name, version])`. the second bit would be injecting that inferred spec in with the rest of the transitive requirement specs at build time before they're all passed to the resolver (and ensuring the dir containing the wheel is injected into the `find-links` paths with precedence). afaict, this would solve for both cases described in 2. but a TODO (ideally w/ github issue link) and iteration is fine also.

Reply:
Cool, thank you for the clarifications and guidance on #3. I'll take a stab and if it proves too complex / time consuming I may just TODO it.

So, related to #1 - the use case for including deps for a `python_dist` is that our internal customer wants to import tensorflow in setup.py so that they can load all the tensorflow includes into include_dirs. Would there be a way around this, if we weren't going to include deps in`python_dist`?

Reply:
> ...our internal customer wants to import tensorflow in setup.py so that they can load all the tensorflow includes into include_dirs. Would there be a way around this, if we weren't going to include deps inpython_dist?

the dependencies declared on this target would never be made available to the executing `setup.py` for use in the current model. they would only be pulled in transitively for any target roots and resolved into deps for a final pex product (pex, chroot or otherwise).

setuptools/`setup.py` et al already make an affordance for this case with `setup_requires`. this ensures the declared `setup_requires` deps are present at *setup.py execution* time vs *install* time. making it so that the user could declare "i need tensorflow during setup.py execution time" would, afaict, involve plumbing full support for `setup_requires` in `SetupPyRunner` via either native `setup.py` declarations or BUILD file declarations. I think from the pants/pex side, this would involve composing a setup-time pex/chroot and executing the `setup.py` inside of that where it'd be able to `import tensorflow` from a hermetic environment.

we typically don't deal with `setup_requires` much in pants/pex land because of the emphasis on bdists (wherein `setup.py` is pre-executed for us) for repeatability - but it'd certainly be a nice feature to have for `python_dist` consumption, since it'd bring us closer to parity with `pip`.